README.md
# vscode-progress-abl-logfile-analizer
A Visual Studio Code extension that analyses Progress ABL log files.


CHANGELOG.md
# Change Log

### 0.0.1 - 02 Aug 2019

* First Commit: **NOT FUNCTIONAL**
  * It is a commit to secure the code, the idea of how its gona work is read.


test/testRunner.ts
'use strict';

import jasmine = require('jasmine');
import * as path from 'path';
import * as util from 'util';

class TestRunner {
    public  run(configPath: string, clb: (error: Error) => void): void {
        const runner = new jasmine(null);
        runner.projectBaseDir = path.dirname(configPath);
        runner.configureDefaultReporter({
            print() {
                const line = util.format.apply(util.format, arguments);
                if (line !== '\n') {
                    console.log(line);
                }
            }
        });
        runner.loadConfigFile(configPath);
        runner.loadSpecs();
        runner.loadHelpers();
        runner.onComplete((passed) => {
            if (passed) {
                clb(null);
            } else {
                clb(new Error('At least one test failed.'));
            }
        });

        runner.execute();
    }
}

// Executed directly
if (!module.parent) {
    if (process.argv.indexOf('-c') === -1) {
        process.stdout.write('The "-c" argument is required. (Path to the jasmine config file)');
        process.exit(1);
    } else if (process.argv.length !== 4) {
        process.stdout.write('The argument is missing. Example to start the test: \
        "node testRunner.js -c ./myConfigPath/jasmine.json"');
        process.exit(1);
    } else {
        let configPath = process.argv[process.argv.indexOf('-c') + 1];
        if (!path.isAbsolute(configPath)) {
            configPath = path.resolve(configPath);
        }
        const runner = new TestRunner();
        runner.run(configPath, (error) => {
            if (error) {
                console.log(error.message);
            } else {
                console.log('All tests succeeded.');
            }
        });
    }
}

export = TestRunner;


test/unittest/TimePeriodCalculator.test.ts
'use strict';

import * as moment from 'moment';
import TimePeriodCalculator = require('../../src/TimePeriodCalculator');

describe('TimePeriodCalculator', () => {
    let testObject: TimePeriodCalculator;
    beforeEach(() => {
        testObject = new TimePeriodCalculator();
    });

    describe('getTimePeriod', () => {

        it('gets the correct timePeriod from "YYYY-MM-DD" and "DD.MM.YYYY".', () => {
            // Arrange
            const logContent = '2018-01-27 [0234ß\n\r0?234%&\n\r} my first logging\
                            \n\r28.01.2020 my second logging 0 $%34&/)(34=}?23\
                            \n\r28.02.2020 my third logging 0 $%34&/)(34=}?23';
            const expected = moment.duration({ days: 1, months: 1, years: 2 });

            // Act
            const result = testObject.getTimePeriod(logContent);

            // Assert
            expect(result.asMilliseconds()).toBe(expected.asMilliseconds());
        });

        it('gets the correct timePeriod from "hh:mm:ss.sss" and "hh:mm:ss.sss".', () => {
            // Arrange
            const logContent = '10:38:28.935 [0234ß\n\b\r0?234%&\n\r} my first logging\
                            \n\r12:50:29,901 my second logging 0 $%34&/)(34=}?23\
                            \n\r16:51:29,001 my third logging 0 $%34&/)(34=}?23';
            const expected = moment.duration({ seconds: 0.066, minutes: 13, hours: 6 });

            // Act
            const result = testObject.getTimePeriod(logContent);

            // Assert
            expect(result.asMilliseconds()).toBe(expected.asMilliseconds());
        });

        it('gets the correct timePeriod from "YYYY-MM-DDThh:mm:ss.sssZ" and "DD/MM/YYYThh:mm:ss,sssZ".', () => {
            // Arrange
            const logContent = '2018-01-27T10:38:28.935Z [0234ß\n\b\r0?234%&\n\r} my first logging\
                            \n\r07.28.2019T16:51:29,001Z my second logging 0 $%34&/)(34=}?23\
                            \n\r2020-02-28T16:51:29.001Z my third logging 0 $%34&/)(34=}?23';
            const expected = moment.duration({ seconds: 0.066, minutes: 13, hours: 6, days: 1, months: 1, years: 2 });

            // Act
            const result = testObject.getTimePeriod(logContent);

            // Assert
            expect(result.asMilliseconds()).toBe(expected.asMilliseconds());
        });

        it('gets the correct timePeriod from "YYYY-MM-DD hh:mm:ss.sss" and "DD/MM/YYY hh:mm:ss,sss".', () => {
            // Arrange
            const logContent = '2018-01-27 10:38:28.935 [0234ß\n\b\r0?234%&\n\r} my first logging\
                            \n\r07.28.2019 16:51:29,001 my second logging 0 $%34&/)(34=}?23\
                            \n\r28/02/2020 16:51:29,001 my third logging 0 $%34&/)(34=}?23';
            const expected = moment.duration({ seconds: 0.066, minutes: 13, hours: 6, days: 1, months: 1, years: 2 });

            // Act
            const result = testObject.getTimePeriod(logContent);

            // Assert
            expect(result.asMilliseconds()).toBe(expected.asMilliseconds());
        });

        it('gets the correct timePeriod from "YYYY-MM-DD hh:mm" and MM/DD/YYYY hh:mm".', () => {
            // Arrange
            const logContent = '2018-01-27 10:38 [0234ß\n\b\r0?234%&\n\r} my first logging\
                            \n\r07.28.2019 16:51:29,001 my second logging 0 $%34&/)(34=}?23\
                            \n\r02/28/2020 16:51 my third logging 0 $%34&/)(34=}?23';
            const expected = moment.duration({ minutes: 13, hours: 6, days: 1, months: 1, years: 2 });

            // Act
            const result = testObject.getTimePeriod(logContent);

            // Assert
            expect(result.asMilliseconds()).toBe(expected.asMilliseconds());
        });

        it('should only consider log statements with the same format (length).', () => {
            // Arrange
            const logContent = '2018-01-27 [0234ß\n\b\r0?234%&\n\r} my first logging\
                            \n\r07.28.2019 16:51:29,001 my second logging 0 $%34&/)(34=}?23\
                            \n\r02/28/2020 my third logging 0 $%34&/)(34=}?23';
            const expected = moment.duration({ days: 1, months: 1, years: 2 });

            // Act
            const result = testObject.getTimePeriod(logContent);

            // Assert
            expect(result.asMilliseconds()).toBe(expected.asMilliseconds());
        });

        it ('should only consider log statements on the very beginning of a line.', () => {
            // Arrange
            const logContent = '2018-01-27 [0234ß\n\b\r0?234%&\n\r} my first logging\
                \n\r07.28.2019 16:51:29,001 my second logging 0 $%34&/)(34=}?23\
                \n\r02/28/2020 my third logging 0 $%34&/)(34=}?23; 2021-03-29 my fourth statement on same line.';
            const expected = moment.duration({ days: 1, months: 1, years: 2 });

            // Act
            const result = testObject.getTimePeriod(logContent);

            // Assert
            expect(result.asMilliseconds()).toBe(expected.asMilliseconds());
        });
    });

    describe('convertToDisplayString', () => {
        const PREFIX = 'Selected: ';
        it('should only consist of "0 ms".', () => {
            // Arrange
            const input = moment.duration({ seconds: 0 });
            const expected = PREFIX + input.asMilliseconds() + 'ms';

            // Act
            const result = testObject.convertToDisplayString(input);

            // Assert
            expect(result).toBe(expected);
        });

        it('should only consist of "ms".', () => {
            // Arrange
            const input = moment.duration({ seconds: 0.123 });
            const expected = PREFIX + input.asMilliseconds() + 'ms';

            // Act
            const result = testObject.convertToDisplayString(input);

            // Assert
            expect(result).toBe(expected);
        });

        it('should only consist of "s" and "ms".', () => {
            // Arrange
            const input = moment.duration({ seconds: 6.123 });
            const expected = PREFIX + input.seconds() + 's, '
                + input.milliseconds() + 'ms';

            // Act
            const result = testObject.convertToDisplayString(input);

            // Assert
            expect(result).toBe(expected);
        });

        it('should only consist of "min", "s" and "ms".', () => {
            // Arrange
            const input = moment.duration({ seconds: 6.123, minutes: 3 });
            const expected = PREFIX + input.minutes() + 'min, '
                + input.seconds() + 's, '
                + input.milliseconds() + 'ms';

            // Act
            const result = testObject.convertToDisplayString(input);

            // Assert
            expect(result).toBe(expected);
        });

        it('should only consist of "h", "min", "s" and "ms".', () => {
            // Arrange
            const input = moment.duration({ seconds: 6.123, minutes: 0, hours: 5 });
            const expected = PREFIX + input.hours() + 'h, '
                + input.minutes() + 'min, '
                + input.seconds() + 's, '
                + input.milliseconds() + 'ms';

            // Act
            const result = testObject.convertToDisplayString(input);

            // Assert
            expect(result).toBe(expected);
        });

        it('should consist of "d", "h", "min", "s" and "ms".', () => {
            // Arrange
            const input = moment.duration({ seconds: 6.123, minutes: 0, hours: 5, days: 15, years: 2 });
            const expected = PREFIX + Math.floor(input.asDays()) + 'd, '
                + input.hours() + 'h, '
                + input.minutes() + 'min, '
                + input.seconds() + 's, '
                + input.milliseconds() + 'ms';

            // Act
            const result = testObject.convertToDisplayString(input);

            // Assert
            expect(result).toBe(expected);
        });
    });
});


test/integrationtest/TimePeriodController.test.ts
'use strict';

import { Position, Selection } from 'vscode';
import * as vscode from 'vscode';
import TimePeriodController = require('../../src/TimePeriodController');

describe('TimePeriodController', () => {

    // Initial situation for all tests:
    // folder 'testLogs' is opened and first file is selected in editor.
    beforeEach((done) => {
        if (vscode.workspace.workspaceFolders !== undefined) {
            const workspaceRootPath = vscode.workspace.workspaceFolders[0].uri.path;
            vscode.workspace.findFiles('*.log').then(
                (uris) => {
                    vscode.workspace.openTextDocument(uris[0]).then(
                        (file) => {
                            vscode.window.showTextDocument(file).then(
                                (editor) => {
                                    editor.selection = new Selection(new Position(0, 0), new Position(0, 0));
                                    done();
                                },
                                (reason) => {
                                    done.fail(reason);
                                });
                        },
                        (reason) => {
                            done.fail(reason);
                        });
                },
                (reason) => {
                    done.fail(reason);
                });
        } else {
            done.fail('No folder was opened!');
        }
    });

    it('should show a status bar item when a range of valid log statements is selcted.', (done) => {

        // Arrange
        const editor = vscode.window.activeTextEditor;
        let controllerSpy = spyOn(TimePeriodController.prototype, 'updateTimePeriod').and.callThrough();

        // Assert
        vscode.window.onDidChangeTextEditorSelection((selectionChangedEvent) => {
            // Only assert once
            if (controllerSpy !== undefined && controllerSpy.calls.count() === 1) {
                expect(controllerSpy.calls.count()).toBe(1);
                expect(typeof (controllerSpy.calls.argsFor(0)[0])).toBe('object');
                expect(controllerSpy.calls.argsFor(0)[0].text).toBe('Selected: 1ms');
                controllerSpy = undefined;
                done();
            }
        });

        // Act
        if (editor) {
            editor.selection = new Selection(new Position(0, 0), new Position(2, 0));
        } else {
            done.fail('No text editor is active!');
        }
    });

    it('should not show a status bar item when a range of invalid log statements is selcted.', (done) => {

        // Arrange
        const editor = vscode.window.activeTextEditor;
        let controllerSpy = spyOn(TimePeriodController.prototype, 'updateTimePeriod').and.callThrough();

        // Assert
        vscode.window.onDidChangeTextEditorSelection((selectionChangedEvent) => {

            // Only check on second call once.
            if (controllerSpy !== undefined && controllerSpy.calls.count() === 2) {
                expect(typeof (controllerSpy.calls.argsFor(1)[0])).toBe('object');
                expect(controllerSpy.calls.argsFor(1)[0].text).toBe('');
                controllerSpy = undefined;
                done();
            }
        });

        // Act
        if (editor) {
            editor.selection = new Selection(new Position(0, 1), new Position(1, 12));
        } else {
            done.fail('No text editor is active!');
        }
    });

    it('should update the status bar item when the active editor switches to another file.', (done) => {

        // Arrange
        let controllerSpy = spyOn(TimePeriodController.prototype, 'updateTimePeriod').and.callThrough();

        // Assert
        vscode.window.onDidChangeActiveTextEditor((selectionChangedEvent) => {

            // Only check on fourth call once.
            if (controllerSpy !== undefined && controllerSpy.calls.count() === 4) {

                // Wait for the TimePeriodController to update the status bar.
                setTimeout(() => {
                    expect(typeof (controllerSpy.calls.argsFor(3)[0])).toBe('object');
                    expect(controllerSpy.calls.argsFor(3)[0].text).toBe('Selected: 1ms');
                    controllerSpy = undefined;
                    done();
                }, 200);
            }
        });

        // Open both files and select some lines then change back to the first file.
        vscode.workspace.findFiles('*.log').then(
            (uris) => {
                // Open and show second file with default selection
                vscode.workspace.openTextDocument(uris[1]).then(
                    (secondFile) => {
                        vscode.window.showTextDocument(secondFile, {
                            preview: false,
                            selection: new Selection(new Position(0, 0), new Position(2, 0))
                        }).then(
                            (editor) => {
                                // Switch back to the first file with default selection.
                                vscode.workspace.openTextDocument(uris[0]).then(
                                    (firstFile) => {
                                        vscode.window.showTextDocument(firstFile, {
                                            preserveFocus: false,
                                            preview: false,
                                            selection: new Selection(new Position(0, 0), new Position(2, 0))
                                        });
                                        done();
                                    },
                                    (reason) => {
                                        done.fail(reason);
                                    });
                            },
                            (reason) => {
                                done.fail(reason);
                            });
                    },
                    (reason) => {
                        done.fail(reason);
                    });
            },
            (reason) => {
                done.fail(reason);
            });
    });
});


src/extension.ts
'use strict';

import * as vscode from 'vscode';
import CustomPatternController = require('./CustomPatternController');
import CustomPatternDecorator = require('./CustomPatternDecorator');
import TimePeriodCalculator = require('./TimePeriodCalculator');
import TimePeriodController = require('./TimePeriodController');
import LogSynthom = require('./logAnalytics');
import LogSynthomController = require('./logAnalyticsController');
import moment = require('moment');

// this method is called when the extension is activated
export function activate(context: vscode.ExtensionContext) {
    // create a new time calculator and controller
    const timeCalculator = new TimePeriodCalculator();
    const timeController = new TimePeriodController(timeCalculator);

    // create log level colorizer and controller
    const customPatternDecorator = new CustomPatternDecorator();
    const customPatternController = new CustomPatternController(customPatternDecorator);

    // create log synthoms and controller
    const logSynthoms = new LogSynthom();
    const logSynthomController = new LogSynthomController(logSynthoms);

    // Add to a list of disposables which are disposed when this extension is deactivated.
    //context.subscriptions.push(timeController, customPatternController, logSynthomController);
}

// this method is called when your extension is deactivated
export function deactivate() {
    // Nothing to do here
}


src/logAnalytics.ts
'use strict';

import { stringify } from "querystring";

// import * as moment from 'moment';
class LogObject {
    public datetimeStart: string;
    public datetimeEnd: string;
    public processId: number;
    public threadId: number;
    public level: number;
    public type: string;
    public message: string;
    public totalTime: string;
    public host: LogObject;
    public childs: LogObject[];
}

class LogSynthom {

    //                                       idThread      tipo
    // [19/07/16@18:41:19.634-0300] P-012080 T-002856 2 AS 4GLTRACE       Run bosau/bosaucontrolprocessrules.p PERSIST [returnTipoBotao - fch/fchsauMD/fchsaucontrolprocessrules.p @ 1302]
    // Datetime                     idProcesso        nivel               message
    public logResult: LogObject[];

    public analytics(logLine: string): any {
        if (logLine.length == 0)
            return;

        let actualHostObject: LogObject = undefined;

        let logGroups = new RegExp('(^\\[[\\d/@:.-]+]) (P-\\d{6}) (T-\\d{6}) (\\d{1}) (\\w+ [\\w ]{14}) (.+)', 'gm');

        let match = logGroups.exec(logLine);
        while (match) {

            let logData: LogObject = new LogObject();
            logData.processId = Number.parseInt(match[2]);
            logData.threadId = Number.parseInt(match[3]);
            logData.level = Number.parseInt(match[4]);
            logData.type = match[5];
            logData.message = match[6];

            let runMatch = new RegExp('(^Run|Func) ([\\w]+) (.+) (\\[.*\\])').exec(logData.message);
            if (runMatch) {
                logData.host = actualHostObject;
                if (actualHostObject != undefined) {
                    actualHostObject.childs.push(logData);
                }
                actualHostObject = logData;
            }

            // arrumar (^Run|Func) ([\\w]+) (.+) (\\[.*\\])
            // nao bate com Run createAuthorization [getAllGuide - rtp/rtdocguia.p @ 1021]

            let returnMatch = new RegExp('(^Return) from (.+) (\\[.*\\])').exec(logData.message);
            if (returnMatch) {
                actualHostObject = logData.host;
            }

            match = logGroups.exec(logLine);
        }
    }

    // Verify if a handle has been deleted
    private verifyHandles(data: string): any {
        let handles: Map<string, string> = new Map<string, string>();

        let handleReg = new RegExp('(Created|Deleted|Deleted-by-GC)+([\\w .-]+)(Handle:)(\\d{4})', 'gm');
        const logLevelRanges = [];

        let match = handleReg.exec(data);
        while (match) {
            if (match[1] == "Created") {
                if (!handles.has(match[4]))
                    handles.set(match[4], "Created");
            } else {
                handles.set(match[4], "Deleted");
            }
            match = handleReg.exec(data);
        }

        let message: string = "";
        handles.forEach((key, val) => {
            if (key == "Created")
                message = "Handle:" + val + " não foi deletado.";
        });
        return message;
    }

    //Separate key events importants to GPS log
    private keyEvents(data: string): any {
        //Return from containsAnyError "no"
    }

}

export = LogSynthom;


src/TimePeriodController.ts
'use strict';

import * as vscode from 'vscode';
import { StatusBarItem } from 'vscode';
import TimePeriodCalculator = require('./TimePeriodCalculator');

class TimePeriodController {

    private _timeCalculator: TimePeriodCalculator;
    private _disposable: vscode.Disposable;
    private _statusBarItem: vscode.StatusBarItem;

    constructor(timeCalculator: TimePeriodCalculator) {
        this._timeCalculator = timeCalculator;

        // Create as needed
        if (!this._statusBarItem) {
            this._statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
        }

        // subscribe to selection change and editor activation events
        const subscriptions: vscode.Disposable[] = [];
        vscode.window.onDidChangeTextEditorSelection(this._onEvent, this, subscriptions);
        vscode.window.onDidChangeActiveTextEditor(this._onEvent, this, subscriptions);

        // update the counter for the current file
        this.updateTimePeriod(this._statusBarItem);

        // create a combined disposable from both event subscriptions
        this._disposable = vscode.Disposable.from(...subscriptions);
    }

    public dispose() {
        this._statusBarItem.dispose();
        this._disposable.dispose();
    }

    public updateTimePeriod(statusBarItem: StatusBarItem) {

        // Get the current text editor
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            this._statusBarItem.hide();
            return;
        }

        const doc = editor.document;

        // Only update status if an log file
        if (doc.languageId === 'log') {

            this._statusBarItem.text = '';
            const timePeriod = this._timeCalculator.getTimePeriod(doc.getText(editor.selection));

            if (timePeriod !== undefined) {

                // Update the status bar
                this._statusBarItem.text = this._timeCalculator.convertToDisplayString(timePeriod);
                this._statusBarItem.show();

            } else {
                this._statusBarItem.hide();
            }

        } else {
            this._statusBarItem.hide();
        }
    }

    private _onEvent() {
        this.updateTimePeriod(this._statusBarItem);
    }
}

export = TimePeriodController;


src/logAnalyticsController.ts
'use strict';

import * as vscode from 'vscode';
import { StatusBarItem } from 'vscode';
import LogSynthom = require('./logAnalytics');

class TimePeriodController {

    private _logSynthom: LogSynthom;
    private _disposable: vscode.Disposable;
    private _statusBarItem: vscode.StatusBarItem;

    constructor(logSynthom: LogSynthom) {
        this._logSynthom = logSynthom;

        // Create as needed
        if (!this._statusBarItem) {
            this._statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
        }

        // subscribe to selection change and editor activation events
        const subscriptions: vscode.Disposable[] = [];
        vscode.window.onDidChangeTextEditorSelection(this._onEvent, this, subscriptions);
        vscode.window.onDidChangeActiveTextEditor(this._onEvent, this, subscriptions);

        // update the counter for the current file
        this.searchSynthom(this._statusBarItem);

        // create a combined disposable from both event subscriptions
        this._disposable = vscode.Disposable.from(...subscriptions);
    }

    public dispose() {
        this._statusBarItem.dispose();
        this._disposable.dispose();
    }

    public searchSynthom(statusBarItem: StatusBarItem) {

        // Get the current text editor
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            this._statusBarItem.hide();
            return;
        }

        const doc = editor.document;

        let text = this._logSynthom.analytics(doc.getText(editor.selection));
        this._statusBarItem.text = text;
        this._statusBarItem.show();




        // // Only update status if an log file
        // if (doc.languageId === 'log') {

        //     this._statusBarItem.text = '';
        //     const timePeriod = this._logSynthom.getTimePeriod(doc.getText(editor.selection));

        //     if (timePeriod !== undefined) {

        //         // Update the status bar
        //         this._statusBarItem.text = this._logSynthom.convertToDisplayString(timePeriod);
        //         this._statusBarItem.show();

        //     } else {
        //         this._statusBarItem.hide();
        //     }

        // } else {
        //     this._statusBarItem.hide();
        // }
    }

    private _onEvent() {
        this.searchSynthom(this._statusBarItem);
    }
}

export = TimePeriodController;


src/TimePeriodCalculator.ts
'use strict';

import * as moment from 'moment';

class TimePeriodCalculator {

    // Converts a given moment.Duration to a string that can be displayed.
    public convertToDisplayString(selectedDuration: moment.Duration): string {
        let text = '';

        if (selectedDuration.asDays() >= 1) {
            text += Math.floor(selectedDuration.asDays()) + 'd';
        }
        if (text !== '') {
            text += ', ' + selectedDuration.hours() + 'h';
        } else if (selectedDuration.hours() !== 0) {
            text += selectedDuration.hours() + 'h';
        }
        if (text !== '') {
            text += ', ' + selectedDuration.minutes() + 'min';
        } else if (selectedDuration.minutes() !== 0) {
            text += selectedDuration.minutes() + 'min';
        }
        if (text !== '') {
            text += ', ' + selectedDuration.seconds() + 's';
        } else if (selectedDuration.seconds() !== 0) {
            text += selectedDuration.seconds() + 's';
        }
        if (text !== '') {
            text += ', ' + selectedDuration.milliseconds() + 'ms';
        } else {
            text += selectedDuration.milliseconds() + 'ms';
        }

        text = 'Selected: ' + text;

        return text;
    }

    public getTimePeriod(data: string): moment.Duration {

        const selContent = data;

        // Clock times with optional timezone ("09:13:16", "09:13:16.323", "09:13:16+01:00", "19:30:47.437+0300")
        let clockPattern = '\\d{2}:\\d{2}(?::\\d{2}(?:[.,]\\d{3,})?)?(?:Z| ?[+-]\\d{2}:\\d{2}| ?[+-]\\d{4})?\\b';

        // ISO dates ("2016-08-23")
        const isoDatePattern = '^\\d{4}-\\d{2}-\\d{2}(?:T|\\b)';

        // Data especifica para log progress ("19/05/22")
        const progressDatePattern = '^[\[]\\d{2}[^\\w\\s]\\d{2}[^\\w\\s]\\d{2}\\b';

        // Culture specific dates ("23/08/2016", "23.08.2016")
        const cultureDatesPattern = '^\\d{2}[^\\w\\s]\\d{2}[^\\w\\s]\\d{4}\\b';

        // Match '2016-08-23 09:13:16.323' as well as '29.01.2018 09:13:34,001'
        const dateTimePattern = '((?:' + isoDatePattern + '|' + cultureDatesPattern + '|' + progressDatePattern + '){1}?[ @]?' +
            '(?:' + clockPattern + '){1})';

        // Only match at the beginning of a line.
        clockPattern = '^' + clockPattern;

        // Match 2017-09-29 as well as 29/01/2019
        const datesPattern = '(' + isoDatePattern + '|' + cultureDatesPattern + '|' + progressDatePattern + '){1}';

        // E.g.: The dateTimePattern ('2016-08-23 09:13:16.323') is preferred
        // over the datesPattern ('2017-09-29 and 29/01/2019')
        const rankedPattern = [dateTimePattern, clockPattern, datesPattern];
        let matches: string[];

        for (const item of rankedPattern) {
            matches = [];
            const timeRegEx = new RegExp(item, 'gm');

            let match = timeRegEx.exec(selContent);

            while (match) {
                matches.push(this._convertToIso(match[0]));
                match = timeRegEx.exec(selContent);
            }

            if (matches.length >= 2) {
                break;
            }
        }

        let timePeriod: moment.Duration;
        timePeriod = undefined;

        if (matches.length >= 2) {
            const firstMoment = moment(matches[0]);
            const lastMoment = moment(matches[matches.length - 1]);

            if (firstMoment.isValid() && lastMoment.isValid()) {

                // used for ISO Dates like '2018-09-29' and '2018-09-29 13:12:11.001'
                timePeriod = moment.duration(lastMoment.diff(firstMoment));
            } else {

                const firstDuration = moment.duration(matches[0]);
                const lastDuration = moment.duration(matches[matches.length - 1]);

                if (moment.isDuration(firstDuration) && moment.isDuration(lastDuration)) {

                    // Used for non ISO dates like '13:12:11.001'
                    timePeriod = moment.duration(lastDuration.asMilliseconds() - firstDuration.asMilliseconds());
                }
            }
        }

        return timePeriod;
    }

    // Converts a given date string to an iso string.
    private _convertToIso(dateString: string): string {

        // [19/12/01 => 2019 'progress abl'
        var re = /(^\[)/;
        if (dateString.match(re)) {
            dateString = dateString.replace('\[', '20');
            dateString = dateString.replace('/', '-');
            dateString = dateString.replace('/', '-');
        }

        // 01.29.2018 or 01/29/2018 => 2018-01-29
        let result = dateString.replace(
            /\b((?:0[1-9])|(?:1[0-2]))[\./-]((?:0[1-9])|(?:[1-2][0-9])|(?:3[0-1]))[\./-](\d{4})/g,
            '$3-$1-$2');

        // 29.01.2018 or 29/01/2018 => 2018-01-29
        result = dateString.replace(
            /\b((?:0[1-9])|(?:[1-2][0-9])|(?:3[0-1]))[\./-]((?:0[1-9])|(?:1[0-2]))[\./-](\d{4})/g,
            '$3-$2-$1');

        // 09:29:02,258 => 09:29:02.258
        result = result.replace(',', '.');

        return result;
    }
}

export = TimePeriodCalculator;
